[✨feat]: RoomPage 레이아웃 및 기본 기능 구현

joyswim opened the pull request:
## 📝 설명
RoomPage 채팅을 제외한 레이아웃 및 기본 기능을 구현했습니다.

### 🔔 추가사항
- store에 `useMeStore`와 `useRoomStore`를 추가했습니다.
  - useMeStore : 내 정보 (id, email, nickname, bojId, profileImage)
  - useRoomStore : 현재 룸에 대한 정보
-   input.ts 인풋의 type에 'number' 추가
- /styles/theme.ts
  - icon size 'XXS' 추가
  - green 진한 색으로 변경
  - FIXED_WIDTH, FIXED_HEIGHT에 CARD 추가

### 페이지 설명
- RoomPage에는 두 가지 Views가 있습니다. RoomViews와 ChatViews입니다.
<img src="https://github.com/1e5i-Shark/algobaro-fe/assets/124686011/32f0a6d4-902c-415b-8b8d-3a313860c8f1" width="500px"/>

- RoomViews는 RoomHeader, MemberList, TestInfo를 가지고 있습니다.
  - RoomHeader는 RoomHeaderInfo와 RoomHeaderButtons를 가지고 있습니다.
  - RoomHeader 내부 컴포넌트에서 발생한 Modal은 RoomHeader에서 띄워줍니다.
<img src="https://github.com/1e5i-Shark/algobaro-fe/assets/124686011/a91750dd-5d91-4751-8736-720df7a0100e" width="600px"/>

### 📷 스크린샷
<img src="https://github.com/1e5i-Shark/algobaro-fe/assets/124686011/396ab085-ef07-49e4-be9e-67b7ace4b420" />

## ✅ PR 유형
- [x] 새로운 기능 추가

## 💻 작업 내용
<!-- PR 본문을 입력하세요. -->
- [x] 방 번호를 클립보드로 복사할 수 있다.
- [x] 방 정보(제목, 태그, 언어, 비밀방)를 보여준다.
- [x] 다크모드, 라이트모드를 바꿀 수 있다.
- [x] 방장은 방에 대한 정보(문제링크, 시간, 비밀번호)를 수정할 수 있다.
- [x] 방장은 수동으로 방장 권한을 위임할 수 있다.
- [x] 방장은 특정 멤버를 강제퇴장할 수 있다.
- [x] 방장은 방의 모든 사람이 '준비'상태일 때 테스트를 시작할 수 있다.
- [x] RoomViews 레이아웃 수정

## 💬 PR 포인트
- 코드 분량이 많습니다. 파일 구조, 주요 기능 위주로 봐주시면 됩니다.
  - 파일 구조, 네이밍이 어렵진 않은지 궁금합니다.
  - 주요 기능은 `pages/RoomPage/RoomView`에 있습니다.
  - 문제 링크 옆 "문제 번호"는 방 정보 수정 후 바뀌었는지 확인하기 위한 용도입니다. 추후 삭제 예정입니다.
- 작업하면서 기능 분리가 어려웠는데 좋은 의견 있으시면 코멘트 부탁드립니다.
- API 연결 관련 코드는 모두 주석 처리되었습니다. 제외하고 PR 진행해주시면 됩니다.
- alert, confirm 창의 커스텀을 하지 않은 상태입니다. 모달 밖을 클릭했을 때 closeModal() 되는 것을 방지할 수 있다면 모달을 활용해 커스텀창을 만들어 볼 수 있을 것 같습니다. ✔ 기존 모달 활용하는 방안으로 채택

joyswim committed:
[🎨design]: 초기 레이아웃 작업

joyswim committed:
Merge branch 'dev' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
Merge branch 'dev' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
[✨feat]: 프로그래밍 언어 로고 상수 선언

joyswim committed:
[✨feat]: 방 정보 데이터 타입 선언

joyswim committed:
[✨feat]: 방 정보를 보여주는 RoomHeader 컴포넌트 생성

joyswim committed:
[✨feat]: RoomHeaderInfo 컴포넌트 생성 초기 레이아웃 잡기

joyswim committed:
[✨feat]: RoomHeaderButtons 컴포넌트 생성 초기 레이아웃 잡기

joyswim committed:
[🚚rename]: RoomHeader 폴더로 이동

joyswim committed:
[✨feat]: 참가자에 대한 UserType 선언

joyswim committed:
[✨feat]: 참가자 컴포넌트 Participants 추가

joyswim committed:
[♻️refactor]: RoomViews의 각 뷰 배럴 export로 내보내기

joyswim committed:
[✨feat]: 테스트 정보에 대한 TestInfo 컴포넌트 추가

joyswim committed:
[♻️refactor]: timer 변수 추가

joyswim committed:
[🎨design]: icon XXS 사이즈 추가

joyswim committed:
[✨feat]: 테스트 정보를 보여주는 TestInfo 컴포넌트 추가

joyswim committed:
Merge branch 'dev' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
Merge branch 'dev' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
[🎨design]: 버튼 글자크기, margin 조정

joyswim committed:
[💄style]: 스타일컴포넌트 ST prefix 붙이기

joyswim committed:
[♻️refactor]: 더미데이터 참가자 데이터 추가

joyswim committed:
[♻️refactor]: UserType image 추가

joyswim committed:
[♻️refactor]: theme WIDTH 상수 CARD 추가

joyswim committed:
[♻️refactor]: 스타일컴포넌트 prefix 코드포맷팅 변경과 데이터 isManger 삭제

joyswim committed:
[🎨design]: padding 조정

joyswim committed:
[🎨design]: ParticipantsContainer 그리드 적용

joyswim committed:
[✨feat]: 참가자 정보를 나타내는 ParticipantsCard 컴포넌트 추가

joyswim committed:
[✨feat]: 참가자 status에 대한 StatusTag 컴포넌트 추가

joyswim committed:
[♻️refactor]: 스타일컴포넌트, 컴포넌트 내보내기 컨벤션에 맞게 수정

joyswim committed:
[🎨design]: Container padding, grow 속성 조정

joyswim committed:
[✨feat]: 채팅 뷰 ChatViews의 ChatInput 레이아웃 구체화

joyswim committed:
[✨feat]: Input 엔터 또는 SendButton 클릭 시 폼 제출하기

joyswim committed:
[♻️refactor]: 테마 WIDTH, HEIGHT에 CARD 추가

joyswim committed:
[🎨design]: CARD의 width, height값 테마값 사용

joyswim committed:
[🎨design]: UI 사이즈 조정

joyswim committed:
Merge branch 'dev' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
[✨feat]: 나가기 버튼 클릭 시 홈으로 이동

joyswim committed:
[🎨design]: Card 컬러 변경

joyswim committed:
[♻️refactor]: 시간, 분 계산과 <tbody> 추가

joyswim committed:
[🎨design]: TestInfo 표 width, padding 조정

joyswim committed:
[♻️refactor]: tags 백엔드 데이터 변경에 따른 배열 변환 로직 추가

joyswim committed:
[♻️refactor]: 룸 관련 API 타입 추가

joyswim committed:
[✨feat]: 개별 방 정보 조회 API 로직 추가

joyswim committed:
Merge branch 'dev' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
[✨feat]: 테스트 시작 시 api 호출과 리다이렉트

joyswim committed:
[✨feat]: 방 번호 복사 기능 추가

joyswim committed:
[💬comment]: Todo 주석 추가

joyswim committed:
[♻️refactor]: 방 정보 api 수정에 대한 타입 및 data 관련 코드 수정

joyswim committed:
[✨feat]: 방 정보 수정 모달 추가

joyswim committed:
[✨feat]: 모달 내부 form 추가

joyswim committed:
[♻️refactor]: 채팅 공백일 때는 onSubmit 처리 안되도록 방어코드 추가

joyswim committed:
[🐛fix]: 중첩 button에 대한 에러 해결

joyswim committed:
[✨feat]: Input의 타입이 number일 때 버튼 숨기기

joyswim committed:
[✨feat]: roomData 스토어 생성 및 컴포넌트 연결

- 일부 변수명 변경

joyswim committed:
[♻️refactor]: 백엔드 데이터 수정에 따른 타입 변경

joyswim committed:
[🚚rename]: 모달 관련 코드 RoomHeader로 이동

joyswim committed:
Merge branch 'dev' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
[🚚rename]: Participants를 백엔드와 동일하게 Members로 변경

joyswim committed:
[🐛fix]: 비밀방 여부에 따라 아이콘 숨기기

joyswim committed:
[🎨design]: 참가자 리스트 화면에 꽉 찼을 때 grid를 변경하는 미디어 쿼리 추가

joyswim committed:
[✨feat]: 모달 취소 시 모달 닫기 추가

joyswim committed:
[♻️refactor]: MemberType 추가

joyswim committed:
[🚚rename]: MemberType 변경에 따른 변수명 변경

joyswim committed:
[✨feat]: 멤버 수가 부족한 경우 MemberCard에 배경색만 보여주기

joyswim committed:
[✨feat]: 문제링크에 대한 정규식 추가

joyswim committed:
[♻️refactor]: 블록 컴포넌트 className 부여와 MemberList 파일명 수정

joyswim committed:
[✨feat]: 나에 대한 정보를 가진 meStore 생성

joyswim committed:
[♻️refactor]: MemberType ready 추가 및 role 수정

joyswim committed:
[🚚rename]: 수정완료 버튼 수정하기로 수정

joyswim committed:
[✨feat]: 내 정보에 대한 MyType 추가

joyswim committed:
[✨feat]: 나가기 버튼 클릭 시 컨펌창 추가

joyswim committed:
[✨feat]: 방 member에서 내 정보를 myRoomData로 추출하여 버튼 분기, 더미데이터 수정

joyswim committed:
Merge branch 'dev' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
[♻️refactor]: Input type number 추가

joyswim committed:
[♻️refactor]: 방장 외 멤버만 케밥 메뉴 보여주기

joyswim committed:
[💄style]: className 변경

joyswim committed:
[💄style]: 스타일컴포넌트 컨벤션 적용

joyswim committed:
[♻️refactor]: 더미데이터 문제플랫폼 BOJ 고정 및 일부 수정

joyswim committed:
[🔥remove]: 안 쓰는 코드 삭제

joyswim committed:
[♻️refactor]: MemberCard props 수정isEmpty 삭제 및 onMenuClick 추가

joyswim committed:
[🎨design]: 비밀방 체크 시 밀리는 현상 수정 width값 조절

joyswim committed:
[♻️refactor]: Room의 API Endpoint 상수 및 타입  추가

joyswim committed:
[♻️refactor]: meStore id null로 초기화

joyswim committed:
[✨feat]: 방 나가기 핸들러 함수 추가

joyswim committed:
[✨feat]: 방장위임, 강제퇴장을 처리하는 handleMenu 함수 추가

joyswim committed:
[♻️refactor]: 호스트만 MemberCard 케밥메뉴 보여주기

joyswim committed:
[♻️refactor]: editRoomProps 추가

joyswim committed:
[🚚rename]: 상수 코드 분리

joyswim committed:
[♻️refactor]: import as 타입 모듈 이름 지정

joyswim committed:
[♻️refactor]: API 함수 분리에 따른 리팩토링

joyswim committed:
[💄style]: 코드 정리(주석, 코드 순서, import문)

joyswim committed:
Merge branch 'dev' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
[🎨design]: 헤더 CSS 간격, 배경색 변경

joyswim committed:
[🎨design]: Tag 배경색 추가

joyswim committed:
[🐛fix]: Input password 눈 모양 안 뜨는 이슈 해결

Co-authored-by: 타래 <tarae7617@naver.com>
Co-authored-by: Cho_Ik_Jun <harry960629@naver.com>

joyswim commented on src/pages/RoomPage/RoomPage.tsx:
```diff
@@ -1,3 +1,62 @@
+    <S.RoomContainer className="room-page">
+      <S.WaitingRoomContainer className="waiting-room">
+        <RoomHeader
+          className="header"
```
익준님: className에 '-container' 추가하기

joyswim commented on src/pages/RoomPage/RoomViews/RoomHeader/ModalRoom.tsx:
```diff
@@ -0,0 +1,161 @@
+    onClose();
+  };
+
+  const allowedKeys = ['Backspace', 'Delete', 'ArrowLeft', 'ArrowRight'];
```
익준님: Tab도 허용하게 하면 좋겠다.

joyswim committed:
[♻️refactor]: 케밥 메뉴 아이콘 hover시 포인터 cursor 바꾸기

joyswim committed:
[💄style]: className 수정

joyswim committed:
[♻️refactor]: Input Tab 허용

harry7435 commented on src/components/MemberCard/StatusTag.tsx:
```diff
@@ -0,0 +1,52 @@
+              : theme.color.red
+        }
+      >
+        {role === 'HOST' ? statusToText[role] : statusToText[role](ready)}
```
p5;
방장과 그냥 멤버의 준비 완료, 대기 중에 대한 조건 분기가 엄청 깔끔하네요...

pySoo commented on src/pages/RoomPage/RoomViews/RoomHeader/RoomHeaderButtons.tsx:
```diff
@@ -0,0 +1,74 @@
+        >
+          방 정보 변경
+        </Button>
+        {theme.mode === 'dark' ? (
+          <Icon
+            className="darkTheme"
+            background={true}
+            onClick={() => {
+              toggleTheme();
+            }}
+            style={{ marginRight: '1rem' }}
+          >
+            <DarkModeRounded fontSize="large" />
+          </Icon>
+        ) : (
+          <Icon
+            className="lightTheme"
+            background={true}
+            onClick={() => toggleTheme()}
+            style={{ marginRight: '1rem' }}
+          >
+            <LightModeRounded fontSize="large" />
+          </Icon>
+        )}
```
요부분 열심히 구현해주셨는데.. 제가 구현하고 말씀을 드릴걸 그랬네요 😭
Common 폴더의 ThemeModeToggleButton이 동일한 기능을 하는 것 같아서
해당 컴포넌트로 사용해보는건 어떠신가요?

joyswim replied:
네 코드도 간결해지고 너무 좋습니다!

pySoo reviewed (approved):
저희 유저들이 가장 자주 보게 될 화면을 굉장히 완성도 있게 구현해주셨네요 🤩
기능도 잘 동작하고 있고 코드도 잘 분리해주신 것 같아요 👍 크으.. 최곱니다
방장 기능, 대기중, 테스트 시작 등등의 상황을 모두 테스트 해볼 수 있게 구현해주셔서 확인하기도 편했습니다!

구두로 나눴던 내용들은 리마인드 용도로 적어두겠습니다~

### 수정 필요 내용

- **`창 확대, 축소시 레이아웃`** 확인
  - 버튼이 눌리지 않는 등 기능상 문제가 없다면 추후 리팩토링 하기
- **`alert, confirm`** 등은 Modal 컴포넌트로 구현
  - 코드 풀이 페이지에서 사용되는 [Confirm 모달](https://github.com/1e5i-Shark/algobaro-fe/blob/d64bfeea59718e2144f3e0a49d49b0549e729db3/src/components/Common/Header/PSHeader.tsx) 참고해보셔도 좋을 것 같아요!

### 소소한 제안

<img src="https://github.com/1e5i-Shark/algobaro-fe/assets/55135881/2b879097-1b4f-4a9d-9a25-14bf86068ad6" width="400" />

- 나중에 채팅이 구현되고 여유가 된다면 ChatView를 접을 수 있도록 구현해보는건 어떠신가요?!
- 가장 간단한 방법으로는 코드 풀이 페이지에서 사용되는 **`창 사이즈 조절 핸들러`** 를 추가해보는 방법이 있을 것 같고, 창을 접을 수 있는 Toggle 버튼을 추가하는 방법도 있을 것 같아요!

harry7435 commented on src/components/MemberCard/StatusTag.tsx:
```diff
@@ -0,0 +1,52 @@
+      </S.StatusText>
+      {role === 'HOST' ? (
+        <Icon color={theme.color.text_primary_color}>
+          <SportsRoundedIcon />
```
방장 아이콘 호루라기 귀엽네요 ㅋㅋㅋ

pySoo replied:
헉 이 아이콘이 호루라기 모양이었군요..? _(저는 어째선지 대포로 착각해버렸네요....😅)_
호옥시 조금 더 직관적인 느낌의 왕관 아이콘을 사용해보는건 어떠신가요?!

joyswim replied:
@pySoo 왕관 아이콘이 MUI에 없어서 호루라기로 대체했습니다! 다른 Icon을 쓸 수 있는 방법이 있을까요?

pySoo replied:
<img width="68" alt="스크린샷 2024-03-11 오전 10 01 03" src="https://github.com/1e5i-Shark/algobaro-fe/assets/55135881/91bea534-856a-4c36-b119-4ee45a6c1f6e">
<img width="71" alt="스크린샷 2024-03-11 오전 10 01 26" src="https://github.com/1e5i-Shark/algobaro-fe/assets/55135881/aba5280c-def5-4e52-ba69-daabae069076">

혹시 이런 훈장 느낌의 뱃지는 어떠신가요?! 크레이지 아케이드 느낌이 나서 가져와봤습니다.. ㅎㅎ

harry7435 commented on src/constants/logos.ts:
```diff
@@ -0,0 +1,10 @@
+export const LOGOS: { [key: string]: string } = {
+  python:
```
```suggestion
  PYTHON:
```
이것도 위와 같이 변경해야 할까요? 서버랑 맞추어야 하겠죠?

joyswim replied:
상수니까 컨벤션 맞춰두겠습니다!

harry7435 commented on src/constants/queryKey.ts:
```diff
@@ -1 +1,2 @@
+export const ROOM_UUID_INFO = 'roomUuidInfo';
 export const MY_INFO_KEY = 'myInfo';
```
```suggestion
export const QUERY_KEY = {
  ROOM: {
    ROOM_UUID_INFO: 'roomUuidInfo',
  },
  MEMBERS: {
    MY_INFO_KEY: 'myInfo',
  },
};
```
쿼리 키도 API 기능 별로 위와 같이 객체로 묶어 놓으면 찾기 편할 것 같네요!
어제 오후에 수현님과 의논할 때 나온 부분인데 같이 의논해보고 어느 기준에서 적용해보면 좋을지 수영님 의견도 궁금합니다!

harry7435 replied:
위와 같이 변경하면 쿼리키를 사용할 때 `QUERY_KEY.ROOM.ROOM_UUID_INFO`와 같이 길어지는 단점은 있으나 자동완성해서 만들기가 편하다는 장점이 있습니다!

joyswim replied:
객체 의견 좋다고 생각해요! 생각해본 적 없었는데 자동완성 시 편할 것 같습니다! 저는 그렇게 진행해보겠습니다~

pySoo replied:
저도 관심사별로 객체로 관리하는 것 너무 좋은 것 같아요!!

pySoo commented on src/components/Common/Menu/MenuText.ts:
```diff
@@ -6,13 +6,7 @@ export interface MenuListProps {
-}
-
-export const MenuText: MenuTextProps = {
+export const MenuText = {
   Report: '신고하기',
   TransferHost: '방장 위임',
   KickOut: '강제 퇴장',
```
오 저도 대문자로 동일하는 것이 컨벤션에 어울린다고 생각합니다!

joyswim replied:
반영됐습니다. Menu 컴포넌트 담당자는 인지만 해주시면 될 것 같습니다!

harry7435 commented on src/pages/RoomPage/RoomViews/TestInfo/TestInfo.tsx:
```diff
@@ -0,0 +1,148 @@
+  const handleClickLink = () => {
+    if (!problemLink) return;
+
+    window.open(problemLink, '_blank');
```
_blank 옵션을 사용할 때는 보안에 취약할 수 있으므로 nooper 파라미터도 추가해주면 좋을 것 같습니다!
[관련 레퍼런스](https://medium.com/@youngminhong/tabnabbing-%EA%B3%B5%EA%B2%A9-%EB%B0%A9%EC%96%B4-%EB%8C%80%EC%B1%85-%EC%A0%95%EB%A6%AC-9276ebf63f94)도 참고해보시면 좋을 거 같아요!
```suggestion
    window.open(problemLink, '_blank', 'noopener');
```

joyswim replied:
보안 취약점에 대해 알려주셔서 감사합니다!

harry7435 commented on src/pages/RoomPage/RoomViews/TestInfo/TestInfo.tsx:
```diff
@@ -0,0 +1,148 @@
+            >
+              <span className="row-content">
+                <h4>문제링크</h4>
+                <Icon className="icon">
+                  <AttachmentRounded />
+                </Icon>
```
p5)
여기도 마찬가지로 커서 속성 주시면 좋을 것 같습니다!
onClick 이벤트를 문제링크란 텍스트가 아닌 Icon 컴포넌트에만 주는 것도 좋겠다는 생각도 드네요!
```suggestion
                <Icon className="icon"
                  style={{
                    cursor: 'pointer',
                  }}
                >
                  <AttachmentRounded />
                </Icon>
```

joyswim replied:
onClick props 추가해서 Icon hover 속성 적용되게 했습니다~
Icon 컴포넌트에만 주게 되면 버튼이 작아지기 때문에 의도적으로 Wrapper에 주었습니다!

harry7435 commented on src/pages/RoomPage/RoomViews/TestInfo/TestInfo.tsx:
```diff
@@ -0,0 +1,148 @@
+          </tr>
+          <tr>
+            <td>
+              <h4>제한시간</h4>
```
p5)
더불어서 제한시간이나 문제링크와 같이 사용자가 복사하거나 드래그가 필요 없는 제목 같은 경우에는
`user-select: none;` 속성을 부여하면 좀 더 깔끔한 느낌을 줄 수 있을 것 같습니다!

harry7435 commented on src/pages/RoomPage/RoomViews/TestInfo/TestInfo.tsx:
```diff
@@ -0,0 +1,148 @@
+              <h4>제한시간</h4>
+            </td>
+            <td>
+              {calcTime.hours > 0 && <span>{calcTime.hours}시간</span>}
```
p5)
시간과 분 사이에 padding이 있으면 좋을 거 같네요!

harry7435 commented on src/pages/RoomPage/RoomViews/TestInfo/TestInfo.tsx:
```diff
@@ -0,0 +1,148 @@
+  });
+
+  const handleStartTest = async () => {
+    await mutation.mutate(`/${roomId}`);
```
p5)
추가적으로 방장 혼자 있을 때 테스트 시작을 누르면 혼자 풀기를 할 것인지 묻는 확인 로직도 있으면 좋을 거 같네요!
```suggestion
    await mutation.mutate(`/${roomId}`);
    if (members.length === 1) {
      // 확인 로직 추가, 모달 or alert
      alert('정말 혼자 풀꺼야?');
    }
```

joyswim replied:
```
const isTestReady = useMemo(() => {
  const result = members.findIndex(member => member.ready === false);
  return result === -1 ? true : false;
}, [members]);
```
해당 로직은 29번줄 isTestReady에서 체크를 하고 있고 "테스트 시작"이 활성화되어 있는 상태라서 인지가 가능할 것 같아 추가하지 않았습니다!

테스트 시작 전 alert가 필요하다면 handleStartTest에서 "바로 테스트를 시작하시겠습니까?"라는 문구로 대체하면 좋을 것 같아요! 5초 뒤 바로 테스트 시작도 괜찮은 것 같습니다.

harry7435 commented on src/services/Room/Room.ts:
```diff
@@ -0,0 +1,51 @@
+  ROOMS_URL,
+} from '@/services/apiEndpoint';
+import { axiosAuthInstance } from '@/services/axiosInstance';
+import * as T from '@/types/room';
+
+// 개별 방 정보 조회
+export const getUuidRoom = async (endPoint: string) => {
+  const response: T.RoomResponse = await axiosAuthInstance.get(
+    `${ROOMS_URL}${endPoint}`
+  );
```
저도 Auth로 인증에 속하는 signIn, signUp을 묶었습니다!
추후 리팩토링해보시죠!

joyswim replied:
api가 많아져서 리팩토링했습니다. 수현님이 알려주신 덕분이죠! 👍

pySoo replied:
오.. 정말 리뷰로 배우는게 많네요! 저도 리팩토링 해봐야겠어유! 👍

harry7435 commented on src/services/Room/Room.ts:
```diff
@@ -0,0 +1,51 @@
+
+// 개별 방 정보 조회
+export const getUuidRoom = async (endPoint: string) => {
+  const response: T.RoomResponse = await axiosAuthInstance.get(
```
p5)
사소한 부분인데 수현님은 axios 메서드의 응답 타입을 제네릭으로 넘겨주시더라구요!
이렇게 하니까 좀 더 깔끔한 느낌이라 저도 적용했었습니다!
수영님도 axios 함수들에 한 번 적용해보시는 거 어떨까요?

```suggestion
  const response = await axiosAuthInstance.get<T.RoomResponse>(
```

harry7435 commented on src/services/Room/Room.ts:
```diff
@@ -0,0 +1,51 @@
+  return response;
+};
+
+// 문제 풀이 시작
+export const startTest = async (endPoint: string) => {
+  return await axiosAuthInstance.post(`${ROOMS_CODES_URL}${endPoint}`);
+};
+
+// 방장 자동 변경
+export const changeHostAuto = async () => {
+  const response = await axiosAuthInstance.get(`${ROOMS_HOST_URL}`);
+  return response;
+};
```
추후 response에 대한 타입 지정도 필요해보입니다!
현재는 api에서 어떤 응답이 올지 애매해서 남겨놓으신 거 같아 unknown으로 뜨네요!

joyswim replied:
api 수정중 상태라 함수 뼈대만 작성해두었습니다! 타입 지정은 api 연결하면서 적용하겠습니다.

harry7435 commented on src/styles/theme.ts:
p5)
세세한 CSS 값까지 신경쓰는 이리오이소 디자인 총책님 🥇

harry7435 commented on src/types/input.ts:
```diff
@@ -11,7 +11,7 @@ export interface InputProps<T extends FieldValues>
   register: UseFormRegister<T>;
   label?: string;
-  type?: 'email' | 'password' | 'text';
+  type?: 'email' | 'password' | 'text' | 'number';
```
p5)
이 부분 예전에 저한테 요청 주셨는데 까먹었어요 ㅠㅠ

joyswim replied:
괜찮습니다!😄

harry7435 commented on src/types/room.ts:
```diff
@@ -0,0 +1,72 @@
+  members: MemberType[];
+}
+
+export interface OmitRoomType
+  extends Omit<RoomType, 'languages' | 'members' | 'problemLink' | 'startAt'> {}
```
동감입니다!! 💯

harry7435 reviewed (approved):
기나긴 코드 리뷰를 드디어 마쳤네요!!
방 내부 페이지를 정말 깔끔하고 세부 동작 및 api 연동까지 자연스럽게 완성된 것 같습니다!!
정말 고생 많으셨습니다 수영님! 💯 

#### 피드백 주요 체크 사항
- `me` Store 네이밍 (https://github.com/1e5i-Shark/algobaro-fe/pull/78#discussion_r1514843121)
- Room 관련 상수명 UpperCase(https://github.com/1e5i-Shark/algobaro-fe/pull/78#discussion_r1514612956)
- 쿼리 키 객체 형태 저장 여부(https://github.com/1e5i-Shark/algobaro-fe/pull/78#discussion_r1514699661)
- 클래스명 접미사 (https://github.com/1e5i-Shark/algobaro-fe/pull/78#discussion_r1514831379)
- early return 문 리팩토링(https://github.com/1e5i-Shark/algobaro-fe/pull/78#discussion_r1514883641)
- 일부 중복 코드 간결화

joyswim committed:
Merge branch 'dev' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
[💄style]: MenuText 상수명 변경

joyswim committed:
[💄style]: LOGOS 상수명 변경 및 toUpperCase 추가

joyswim committed:
[💄style]: 쿼리키 자동완성을 위해 객체화

joyswim commented on src/pages/RoomPage/ChatViews/ChatInput.tsx:
```diff
@@ -0,0 +1,64 @@
+    handleSubmit(onSubmit)();
+  };
+
+  const handleKeyDown = (e: React.KeyboardEvent) => {
+    if (e.key !== 'Enter') {
+      return;
+    } else {
+      handleSubmit(onSubmit)();
+    }
+  };
```
공유 감사합니다! ChatInput은 RoomPage에서 Chat 컴포넌트로 교체될 예정으로 
해당 내용은 Chat 브랜치에서 반영하겠습니다~

joyswim committed:
[💄style]: className container 접미사 추가 및 수정

joyswim commented on src/pages/RoomPage/RoomPage.consts.ts:
```diff
@@ -0,0 +1,16 @@
+import { AccessType, RoleType, RoomStatusType } from '@/types/room';
+
+export const ROOM_STATUS: { [key: string]: RoomStatusType } = {
+  RECRUITING: 'RECRUITING',
+  RUNNING: 'RUNNING',
+};
+
+export const ROOM_ACCESS: { [key: string]: AccessType } = {
+  PUBLIC: 'PUBLIC',
+  PRIVATE: 'PRIVATE',
+};
+
+export const ROOM_ROLE: { [key: string]: RoleType } = {
+  HOST: 'HOST',
+  MEMBER: 'MEMBER',
+};
```
네 좋습니다!

joyswim committed:
[♻️refactor]: myRoomData 함수 의존성 배열에 roomId 삭제하여 불필요한 렌더링 방지

joyswim commented on src/pages/RoomPage/RoomPage.tsx:
```diff
@@ -1,3 +1,62 @@
+  const { roomData } = useRoomStore();
+  const { members, roomId } = roomData;
+
+  const myRoomData = useMemo(
+    () => members.filter(member => member.id === me.id)[0],
+    [roomId, members]
+  );
```
디테일 챙겨주셔서 감사합니다~
저는 새로운 방에 들어갔을 때를 생각해서 roomId를 넣었는데 새로운 방에 들어갔을 때 RoomPage가 렌더링되면서 useRoomData가 다시 계산되겠네요!

joyswim commented on src/pages/RoomPage/RoomPage.tsx:
```diff
@@ -1,3 +1,62 @@
+
+  return (
+    <S.RoomContainer className="roompage">
+      <S.WaitingRoomContainer className="waitingroom">
```
roompage는 root 바로 밑에 있을 것 같아서 roompage 그대로 두었습니다.
이 부분도 container가 필요하다고 생각하시는 걸까요?

harry7435 replied:
솔직히 컨벤션적인 문제고 필수적인 것은 아니라서 추후에 리팩토링할 때 체크해보면 좋을 것 같습니다.

joyswim committed:
[♻️refactor]: api 응답 수정에 따라 response 변경

joyswim commented on src/pages/RoomPage/RoomViews/MemberList/MemberList.tsx:
```diff
@@ -0,0 +1,107 @@
+  const mutation = useMutation({
+    mutationFn: changeHost,
+    onSuccess: response => {
+      const { organizerId } = response;
```
해당 부분 수정된 내용인데 놓칠 뻔했네요. api까지 꼼꼼히 확인해주셔서 감사합니다!

joyswim commented on src/pages/RoomPage/RoomViews/MemberList/MemberList.tsx:
```diff
@@ -0,0 +1,107 @@
+      const { organizerId } = response;
+      // Todo: API 테스트
+      const updatedMembers: MemberType[] = members.map(member => {
+        if (member.id === organizerId) {
+          return { ...member, role: ROOM_ROLE.HOST };
+        } else if (member.role === ROOM_ROLE.HOST) {
+          return { ...member, role: ROOM_ROLE.MEMBER };
+        } else {
+          return member;
+        }
```
코드 디테일 너무 좋습니다!

joyswim committed:
[💄style]: elseif문 if문으로 수정

joyswim commented on src/pages/RoomPage/RoomViews/MemberList/MemberList.tsx:
```diff
@@ -0,0 +1,107 @@
+
+      setRoomData({ members: updatedMembers });
+    },
+  });
```
너무 좋은 방안이라고 생각합니다! 추후 리팩토링 사항으로 남겨두겠습니다!

joyswim committed:
[💄style]: elseif문 if문으로 수정

joyswim committed:
Merge branch 'feature/#32-Room' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
[🚚rename]: renderMembers 변수명 memberCards로 수정

joyswim commented on src/pages/RoomPage/RoomViews/MemberList/MemberList.tsx:
```diff
@@ -0,0 +1,107 @@
+    }
+  };
+
+  const renderCards = useMemo(() => {
```
좋습니다~

joyswim committed:
[♻️refactor]: key값 uuid 수정

joyswim commented on src/pages/RoomPage/RoomViews/RoomHeader/ModalRoom.tsx:
```diff
@@ -0,0 +1,161 @@
+    const newPassword = () => {
+      const newAccessType: AccessType = isPrivate ? 'PRIVATE' : 'PUBLIC';
+
+      if (isPrivate) {
+        return {
+          password: data.password,
+          roomAccessType: newAccessType,
+        };
+      } else {
+        return {
+          password: '',
+          roomAccessType: newAccessType,
+        };
+      }
```
삼항연산자로 훨씬 가독성이 좋아졌네요. 감사합니다~

joyswim commented on src/pages/RoomPage/RoomViews/RoomHeader/ModalRoom.tsx:
```diff
@@ -0,0 +1,161 @@
+            />
+          )}
+        </S.PasswordWrapper>
+      </S.FormWrapper>
+      <S.ModalButtonsWrapper>
+        <Button
+          type="submit"
+          width="30rem"
+          onClick={handleComplete}
+        >
+          수정하기
+        </Button>
+        <Button
+          backgroundColor={theme.color.transparent_50}
+          onClick={onClose}
+        >
+          취소
+        </Button>
+      </S.ModalButtonsWrapper>
```
꼼꼼한 리뷰 감사합니다~ 해당 부분 반영했습니다.

joyswim committed:
[♻️refactor]: modalRoom PR 반영

- mutation 주석처리
- elseif문 if문으로 리팩토링
- 반복코드 validateKeys 함수 분리
- ModalButtonsWrapper 위치 수정

joyswim committed:
[♻️refactor]: 테마 토글버튼 공통컴포넌트로 교체

joyswim commented on src/pages/RoomPage/RoomViews/RoomHeader/RoomHeaderInfo.tsx:
```diff
@@ -0,0 +1,77 @@
+  const { roomData } = useRoomStore();
+  const { roomUUID, title, tags, languages, roomAccessType } = roomData;
+
+  const handleCopyRoomId = () => {
+    window.navigator.clipboard.writeText(roomUUID);
+
+    alert('방 번호가 복사되었습니다!');
+  };
```
수현님은 새로운 기능을 보면 canIuse를 찾아보나요? 이런 정보를 잘 알고 있는 것 같아 궁금합니다!

joyswim replied:
execCommand는 지원 중단 되었기 때문에 방어코드만 넣어두겠습니다!

pySoo replied:
헉 Deprecated 되었군요 😭 확인 감사합니다!
저는 QA로 맞으면서 습득한 편입니다... ㅋㅋㅋ web API의 경우는 canIuse를 매번 찾아보는 편은 아니지만,,

라이브러리 같은 경우에는 새로운 버전이면 최신 브라우저에서만 대응이 되거나 호환이 되지 않는 문제가 있어서
공식 문서를 보고 버전을 맞추는 편이에요!
(react-query 5는 21년에 출시된 safari 15버전부터 지원한다고 합니다.. 덜덜)

joyswim commented on src/pages/RoomPage/RoomViews/RoomHeader/ModalRoom.tsx:
```diff
@@ -0,0 +1,161 @@
+    if (!regex.test(event.key) && !allowedKeys.includes(event.key)) {
+      event.preventDefault();
+    }
+  };
```
커스텀 훅으로 분리하지는 않고 validateKeys 함수로 분리했습니다!

joyswim committed:
[♻️refactor]: 복사하기 기능 방어코드 추가

joyswim commented on src/pages/RoomPage/RoomPage.tsx:
```diff
@@ -1,3 +1,62 @@
+import { MemberList, RoomHeader, TestInfo } from './RoomViews';
+
 export default function RoomPage() {
-  return <div>RoomPage</div>;
+  const { me } = useMeStore();
```
스토어명 useMyInfoStore(), 정보에 대한 상태는 myInfo 이렇게 말씀하시는 걸까요?

harry7435 replied:
네! 전 그렇게 생각했습니다!

pySoo replied:
myInfo라는 이름 좋네유!

joyswim replied:
반영해두도록 하겠습니다~

joyswim commented on src/pages/RoomPage/RoomViews/RoomHeader/RoomHeaderInfo.tsx:
```diff
@@ -0,0 +1,77 @@
+      <S.RoomIdWrapper className="roomId">
+        <S.CopyRoomIdTag onClick={handleCopyRoomId}>
+          <S.TextId>방 번호 복사</S.TextId>
+          <Icon size="XXS">
+            <ContentCopyRounded />
+          </Icon>
```
onClick props 추가해서 cursor: pointer 적용되게 하겠습니다!

joyswim committed:
[🎨design]: 복사하기 아이콘 onClick props 추가

joyswim commented on src/pages/RoomPage/RoomViews/RoomHeader/RoomHeaderInfo.tsx:
```diff
@@ -0,0 +1,77 @@
+        )}
+        <h2>{title}</h2>
+      </S.TitleWrapper>
+      <S.TagsWrapper className="tagsGroup">
+        {tags.map((tag, index) => (
+          <S.TagWrapper key={tag + index}>
+            <Tag
+              height="2.6rem"
+              fontSize="1.4rem"
+              mode="normal"
+              backgroundColor={theme.color.transparent_50}
+              tagId={tag + index}
+            >
+              {tag}
+            </Tag>
+          </S.TagWrapper>
+        ))}
+      </S.TagsWrapper>
```
저도 index는 지양해야 한다고 생각합니다!
참고로 uuid의 단점은 매 렌더링마다 새로운 uuid가 할당되어서 기존과 달라지지 않아도 다시 리렌더링을 한다고 합니다. 따라서 uuid는 index의 대체재이지만 key의 목적과는 다르게 쓰일 수 있다고 하네요! 아래 관련 블로그 글을 남겨뒀습니다.

[React에서 UUID를 key로 쓸 때 주의할 점](https://zungta.com/posts/8fd32f52-24aa-4707-bdfb-15bd0741d2eb)

pySoo replied:
헉.. 그렇군요 고유값을 만들어주는 특성이
오히려 렌더링에 영향을 줄 수도 있겠네요..! 새로 배워갑니다 자료 공유 넘 감사해요!! 👍

joyswim committed:
[♻️refactor]: 태그 key값 uuid로 수정

joyswim commented on src/pages/RoomPage/RoomViews/TestInfo/TestInfo.tsx:
```diff
@@ -0,0 +1,148 @@
+    return { hours, minutes };
+  }, [timeLimit]);
+
+  const changeMemberData = (newData: Partial<MemberType>) => {
```
@harry7435 
유틸리티 타입을 배울 때 이걸로 타입을 재사용할 수 있겠다고 생각하고 있었습니다. 
제대로 사용하기 시작한 건 타입을 선언하면서 사소하게 타입이 변경되는 경우가 있었고 그 과정에서 타입이 일치하지 않는 실수가 생기더라구요.
그래서 아래 같이 잘 정리된 문서 보고 간단하게 표로 정리해두고 참고하고 있습니다! 자주 사용되는 게 Partial, Pick, Omit인 것 같아요.

[참고](https://typescript-kr.github.io/pages/utility-types.html#recordkt)

joyswim committed:
[♻️refactor]: myIndex를 찾지 못할 경우 방어코드 추가

joyswim commented on src/pages/RoomPage/RoomViews/TestInfo/TestInfo.tsx:
```diff
@@ -0,0 +1,148 @@
+  }, [timeLimit]);
+
+  const changeMemberData = (newData: Partial<MemberType>) => {
+    const myIndex = members.findIndex(member => member.id === me.id);
+
```
꼼꼼한 수현님 방어코드가 일상이 될 때까지 맘껏 리뷰 부탁드립니다~

pySoo replied:
😁👍👍👍
(그렇게 100개에 도달해버린 댓글 수... ㅎㅎ)

joyswim commented on src/pages/RoomPage/RoomViews/TestInfo/TestInfo.tsx:
```diff
@@ -0,0 +1,148 @@
+  const mutation = useMutation({
+    mutationFn: startTest,
+    onError: () => {
+      alert('테스트를 시작하지 못했습니다.');
```
어떤 문구를 해야 할지 고민했는데 좋은 의견입니다!

joyswim committed:
[♻️refactor]: 테스트 시작 실패 시 알림 문구 수정

joyswim committed:
[♻️refactor]: mutation 반환 값 mutation에서 객체 구조 분해 할당으로 수정

joyswim committed:
[♻️refactor]: tabnabbing 보안을 위해 window.open() noopener 파라미터 추가

joyswim committed:
[♻️refactor]: Icon hover 속성 적용을 위해 onClick 추가

joyswim committed:
[💄style]: 시간, 분 사이 공백 추가

joyswim committed:
[♻️refactor]: axios 함수 응답 타입 제네릭 변경

joyswim committed:
[♻️refactor]: API 수정에 따른 타입 및 데이터 수정

joyswim committed:
[♻️refactor]: ValidateEnter API 추가

joyswim committed:
[🚚rename]: 멤버의 id, memberId로 수정

joyswim committed:
[🚚rename]: 멤버의 id, memberId로 수정

joyswim committed:
Merge branch 'feature/#32-Room' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
[♻️refactor]: status 변환하는 객체를 함수로 리팩토링 및 role 상수 교체

joyswim committed:
[🚚rename]: API 변경에 따른 일부 타입 수정 및 적용

joyswim committed:
[♻️refactor]: 방장만 방 정보 수정하기 버튼 활성화

joyswim committed:
[♻️refactor]: 방 정보 수정 요청 필수 속성 반영

joyswim committed:
[💄style]: Store 네이밍 통일

joyswim committed:
[🚚rename]: myInfo 네이밍 통일

joyswim committed:
[🎨design]: 최상위 컴포넌트 height 조정 및 TestInfo 부분 선택 안되는 기능 추가

joyswim committed:
[✨feat]: 방 수정 API 연결

joyswim committed:
[💄style]: 버튼 텍스트 수정

joyswim committed:
[♻️refactor]: API 수정으로 사용되지 않는 API 제거

joyswim committed:
[♻️refactor]: 쿼리 훅 분리

joyswim committed:
[♻️refactor]: 별도의 import문 객체 구조 분해 할당으로 가져오기 수정

joyswim committed:
[💄style]: 주석 삭제 및 불필요한 텍스트 삭제

joyswim committed:
[♻️refactor]: 1차 배포를 위한 주요 컴포넌트 리팩토링 및 파일 이동

joyswim committed:
[♻️refactor]: API 연결 및 입장 로직 추가

joyswim committed:
[♻️refactor]: MyInfoStore, RoomStore 더미데이터 추가

joyswim committed:
[💄style]: 모든 방 멤버 준비상태 true로 변경

joyswim committed:
[💄style]: createRoom API 엔드포인트 수정

joyswim committed:
Merge branch 'dev' of https://github.com/1e5i-Shark/algobaro-fe into feature/#32-Room

joyswim committed:
[🔥remove]: useRooms 코드 이동
